### ISSUE TITLE
Per-replication callbacks?

### ISSUE THREAD

**Issue by ndouglas** [couchbase/couchbase-lite-ios#330]
(Possibly related to #293, #78, #79)

**TL;DR**: I'm interested in a per-replication callback.  I understand there's already -[CBLReplication propertiesTransformationBlock], but I'm concerned that:
- it might not be the best place for a long asynchronous operation like a network operation
- if there is an error, it doesn't seem possible to reject the revision

Basically, I need to implement uploading to S3 and a couple other different things, but there doesn't seem to be a good place to accomplish this.  I figure this might also be good for people who want per-peer document validation before something enters the database, or at least per-peer recovery from those issues.

Normally I'd just implement it myself, but since CBLReplication quickly turns into CBL_Replicator, I figured it was prudent in this case to consult you :-)
